javascript/node/index.js

```diff
@@ -0,0 +1,111 @@
+import {
+    createRequest,
+    isEmptyMessage,
+    streamBlocks,
+    createAuthInterceptor,
+    createRegistry,
+    fetchSubstream
+} from '@substreams/core';
+import { createConnectTransport } from "@connectrpc/connect-node";
+import fs from "fs";
+
+const TOKEN = process.env.SUBSTREAMS_API_TOKEN
+const ENDPOINT = "https://mainnet.eth.streamingfast.io"
+const SPKG = "https://storage.googleapis.com/substreams-registry/spkg/ethereum-explorer-v0.1.1.spkg"
+const MODULE = "map_block_meta"
+const CURSOR_FILE = "cursor"
+const START_BLOCK = '100000'
+const STOP_BLOCK = '+10000'
+
+const fetchPackage = async () => {
+    return await fetchSubstream(SPKG)
+}
+
+const getCursor = async () => {
+    return fs.promises.readFile(CURSOR_FILE)
+}
+
+// In this example, the cursor is persisted in a file.
+const writeCursor = async cursor => {
+    try {
+        await fs.promises.writeFile(CURSOR_FILE, cursor)
+    } catch (e) {
+        throw new Error("COULD_NOT_COMMIT_CURSOR")
+    }
+}
+
+const unpackAndCommitCursor = async (response, registry) => {
+  if (response.message.case === "blockScopedData") {
+    const output = response.message.value.output?.mapOutput;
+    const cursor = response.message.value.cursor;
+
+    await writeCursor(cursor);
+
+    if (output !== undefined) {
+      const message = output.unpack(registry);
+      if (message === undefined) {
+        throw new Error(`Failed to unpack output of type ${output.typeUrl}`);
+      }
+
+      return message;
+    }
```

Comment:
So this is a potentially a problem because the `cursor` is written and then the message is processed. 

If the process crashes after the unpack and commit has been called but before the app correctly for example save to the db, on restart you would miss a block.

The cursor saving must always be done after the app processed the block. If we want to have it one function, the "processor" should be received as an argument and cursor saved after it has successfully return.

```js
const processMessage = async (response, registry, processor) => {
   // unpack()
   processor(message)
   // write cursor
```

With this flow, we write the cursor only if the processor succeeded. And in the eventuality that the cursor writing fails, the processor will re-do the same block but won't miss anything.